Playwright: add unit tests for calypso-e2e module

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR partially implements #53033.


#### Testing instructions

1. check out the branch.

either:
2. navigate to `packages/calypso-e2e`
3. run `yarn jest`

or
2. run `yarn test-packages`

Note, that if you have `NODE_CONFIG_ENV` environment variable set, tests will trigger a warning from `config` that it could not find any configuration. This is expected and should not matter since I ended up mocking the call anyway. If this should be suppressed, I am sure there is a way to do so with a mock, but in the TeamCity environment the aforementioned environment variable isn't set anyway so the warning doesn't even show up.

Related to #53033.

worldomonation committed:
Unit tests for @automattic/calypso-e2e BrowserManager

worldomonation committed:
Add MediaHelper tests.

worldomonation committed:
Implement additional tests

worldomonation committed:
Additional tests for getScreenshotDir

worldomonation committed:
Add tests for getVideoDir()

worldomonation committed:
Move all describe blocks into one top level block to remove redundant beforeAll and afterAll setups.

scinos commented on packages/calypso-e2e/package.json:
```diff
@@ -21,10 +21,18 @@
+		"build": "tsc --build ./tsconfig.json",
+		"test": "jest"
+	},
+	"jest": {
```
Packages in the monorepo are expected to implement tests by having a `jest.config.js` file that extends from `../../test/packages/jest-preset.js` (plus any needed customization).

They also don't need to depend on `jest` or implement a script `test`, as testing is run using the scripts `yarn test-packages ...` declared in the root of the project.

worldomonation replied:
I admit I was not aware of this, I will look for other examples under `packages` and follow them.

scinos replied:
Probably is not documented anywhere 😬

scinos commented on packages/calypso-e2e/test/browser-helper.test.ts:
```diff
@@ -0,0 +1,110 @@
+/**
+ * @file Tests the BrowserHelper functions using process.env and stubbed config file.
+ * @author Edwin Takahashi
```
We usually don't have these header in files. It doesn't add more info to what `git blame` tells us.

worldomonation replied:
Will be addressed in next patch, and the removal will include another instance of this header.

scinos commented on packages/calypso-e2e/package.json:
```diff
@@ -21,10 +21,18 @@
+		"@types/jest": "^25.2.3",
+		"@types/node": "^15.0.2",
+		"jest": "^26.6.3",
+		"ts-jest": "^26.5.6"
```
Why `ts-jest` instead of `@babel/preset-typescript`?

Looks like `ts-jest` also gives us type checking, but we already have a CI job for that (as long as your TS package is declared in `./packages/tsconfig.json`)

worldomonation replied:
I followed a tutorial I found that suggested using `ts-jest`. No deeper reason than that, so not hung up on sticking with `ts-jest`. I will try getting this to work with `@babel/preset-typescript`.

worldomonation committed:
Migrate `@automattic/calypso-e2e` unit tests to use the inherited Jest configuration.
- remove dependencies on Jest from package.json.
- remove test config values, with it being replaced with a mocked `config` module.
- remove file header.

worldomonation committed:
Changes to package.json

kwight commented:
I tested with both commands; `media-helper.test.js` and `browser-helper.test.js` both came back as passing (including the config-related console errors you predicted when testing from within the package).

fullofcaffeine reviewed (approved):
Nice! I must say I rarely see tests for test helpers. This is refreshing! A neat side-effect is the executable docs these functions have now.

scinos commented on packages/calypso-e2e/test/browser-helper.test.ts:
```diff
@@ -0,0 +1,105 @@
+	env = process.env;
+} );
+
+describe( 'BrowserHelper Tests', function () {
```
This external `describe` is not doing anything, I'd consider deleting it.

Or moving the `beforeAll` and `afterAll` hooks inside.

worldomonation replied:
I had thought this external hook was acting to limit the number of times the hooks run to 1 per file. 

With that said, I will try moving the hooks.

scinos commented on packages/calypso-e2e/test/browser-helper.test.ts:
```diff
@@ -0,0 +1,105 @@
+
+describe( 'BrowserHelper Tests', function () {
+	describe( `Test: getViewportName`, function () {
+		test.each`
```
Personally I prefer to use the array API (https://jestjs.io/docs/api#1-testeachtablename-fn-timeout). The table API suggest the indentation of the columns is important and people may waste time formatting it. The array API doesn't have this problem as `prettier` will auto-format it.

That being said, I really like the `each` approach, combining good and bad inputs!

worldomonation replied:
Interestingly, I didn't spend any time formatting this! `wp-prettier` seems to handle the table format as well.
